python/demo/demo_pml/demo_pml.py

```diff
@@ -0,0 +1,606 @@
+
+from dolfinx.io import VTXWriter, gmshio
+from efficiencies_pml_demo import calculate_analytical_efficiencies
+from mesh_wire_pml import generate_mesh_wire
+from mpi4py import MPI
+from petsc4py import PETSc
+
+import ufl
+from dolfinx import fem, mesh, plot
+
+# -
+
+# Since we want to solve time-harmonic Maxwell's equation, we require that the demo is
+# executed with DOLFINx (PETSc) complex mode.
+
+if not np.issubdtype(PETSc.ScalarType, np.complexfloating):
+    print("Demo should only be executed with DOLFINx complex mode")
+    exit(0)
+
+
+# WNow, let's consider an infinite metallic wire immersed in
+# a background medium (e.g. vacuum or water). Let's now
+# consider the plane cutting the wire perpendicularly to
+# its axis at a generic point. Such plane $\Omega=\Omega_{m}
+# \cup\Omega_{b}$ is formed by the cross-section
+# of the wire $\Omega_m$ and the background medium
+# $\Omega_{b}$ surrounding the wire. We limit the background medium
+# with a squared perfectly matched layer (or shortly PML), which will act as an absorber for
+# outgoing scattered waves.
+#
+# The goal of this demo is to calculate the electric field $\mathbf{E}_s$
+# scattered by the wire when a background wave $\mathbf{E}_b$
+# impinges on it. We will consider a background plane wave at
+# $\lambda_0$ wavelength, which can be written analytically as:
+#
+# $$
+# \mathbf{E}_b = \exp(\mathbf{k}\cdot\mathbf{r})\hat{\mathbf{u}}_p
+# $$
+#
+# with $\mathbf{k} = \frac{2\pi}{\lambda_0}n_b\hat{\mathbf{u}}_k$
+# being the wavevector of the
+# plane wave, pointing along the propagation direction,
+# with $\hat{\mathbf{u}}_p$ being the
+# polarization direction, and with $\mathbf{r}$ being a
+# point in $\Omega$.
+# We will only consider $\hat{\mathbf{u}}_k$ and $\hat{\mathbf{u}}_p$
+# with components belonging
+# to the $\Omega$ domain and perpendicular to each other,
+# i.e. $\hat{\mathbf{u}}_k \perp \hat{\mathbf{u}}_p$
+# (transversality condition of plane waves).
+# Using a Cartesian coordinate system for $\Omega$,
+# and by defining $k_x = n_bk_0\cos\theta$ and
+# $k_y = n_bk_0\sin\theta$, with $\theta$ being the angle
+# defined by the propagation direction $\hat{\mathbf{u}}_k$
+# and the horizontal axis $\hat{\mathbf{u}}_x$,
+# we have:
+#
+# $$
+# \mathbf{E}_b = -\sin\theta e^{j (k_xx+k_yy)}\hat{\mathbf{u}}_x
+# + \cos\theta e^{j (k_xx+k_yy)}\hat{\mathbf{u}}_y
+# $$
+#
+# The function `background_field` below implements this analytical
+# formula:
+
+# +
+
+def background_field(theta: float, n_b: float, k0: complex,
+                     x: np.typing.NDArray[np.float64]):
+
+    kx = n_b * k0 * np.cos(theta)
+    ky = n_b * k0 * np.sin(theta)
+    phi = kx * x[0] + ky * x[1]
+    return (-np.sin(theta) * np.exp(1j * phi), np.cos(theta) * np.exp(1j * phi))
+
+
+# -
+
+# For convenience, we define the $\nabla\times$ operator for a 2D vector
+
+# +
+
+def curl_2d(a: fem.Function):
+    return ufl.as_vector((0, 0, a[1].dx(0) - a[0].dx(1)))
```

Comment:
Can't you use `ufl.curl` instead of defining your own 2D version of `curl`?

Reply:
I think that using `ufl.curl` in this case is not possible since the problem is two-dimensional, and `ufl.curl` would try to evaluate the derivative of `Es` along the third dimension, resulting in an error. Therefore we need to define this version of the $\nabla \times$ operator.

Reply:
At least for me it works in a 2D problem.

Reply:
To be clearer, if I use this weak form:

```python
F = - ufl.inner(ufl.curl(Es_3d), ufl.curl(v_3d)) * dDom \
    + eps * (k0**2) * ufl.inner(Es, v) * dDom \
    + (k0**2) * (eps - eps_bkg) * ufl.inner(Eb, v) * dDom \
    - ufl.inner(ufl.inv(mu_x) * ufl.curl(Es_3d), ufl.curl(v_3d)) * dPml_x \
    - ufl.inner(ufl.inv(mu_y) * ufl.curl(Es_3d), ufl.curl(v_3d)) * dPml_y \
    - ufl.inner(ufl.inv(mu_xy) * ufl.curl(Es_3d), ufl.curl(v_3d)) * dPml_xy \
    + (k0**2) * ufl.inner(eps_x * Es_3d, v_3d) * dPml_x \
    + (k0**2) * ufl.inner(eps_y * Es_3d, v_3d) * dPml_y \
    + (k0**2) * ufl.inner(eps_xy * Es_3d, v_3d) * dPml_xy
```

I get this error:

```bash
Traceback (most recent call last):
  File "/root/dolfinx/python/demo/demo_pml/demo_pml.py", line 444, in <module>
    a, L = ufl.lhs(F), ufl.rhs(F)
  File "/root/ufl/ufl/formoperators.py", line 69, in lhs
    form = expand_derivatives(form)
  File "/root/ufl/ufl/algorithms/ad.py", line 25, in expand_derivatives
    form = apply_algebra_lowering(form)
  File "/root/ufl/ufl/algorithms/apply_algebra_lowering.py", line 175, in apply_algebra_lowering
    return map_integrand_dags(LowerCompoundAlgebra(), expr)
  File "/root/ufl/ufl/algorithms/map_integrands.py", line 46, in map_integrand_dags
    return map_integrands(lambda expr: map_expr_dag(function, expr, compress),
  File "/root/ufl/ufl/algorithms/map_integrands.py", line 27, in map_integrands
    mapped_integrals = [map_integrands(function, itg, only_integral_type)
  File "/root/ufl/ufl/algorithms/map_integrands.py", line 27, in <listcomp>
    mapped_integrals = [map_integrands(function, itg, only_integral_type)
  File "/root/ufl/ufl/algorithms/map_integrands.py", line 35, in map_integrands
    return itg.reconstruct(function(itg.integrand()))
  File "/root/ufl/ufl/algorithms/map_integrands.py", line 46, in <lambda>
    return map_integrands(lambda expr: map_expr_dag(function, expr, compress),
  File "/root/ufl/ufl/corealg/map_dag.py", line 36, in map_expr_dag
    result, = map_expr_dags(function, [expression], compress=compress,
  File "/root/ufl/ufl/corealg/map_dag.py", line 99, in map_expr_dags
    r = handlers[v._ufl_typecode_](v, *[vcache[u] for u in v.ufl_operands])
  File "/root/ufl/ufl/algorithms/apply_algebra_lowering.py", line 168, in curl
    return as_vector((c(1, 2), c(2, 0), c(0, 1)))
  File "/root/ufl/ufl/algorithms/apply_algebra_lowering.py", line 161, in c
    return a[j].dx(i) - a[i].dx(j)
  File "/root/ufl/ufl/exproperators.py", line 498, in _dx
    d = Grad(d)
  File "/root/ufl/ufl/differentiation.py", line 140, in __new__
    dim = find_geometric_dimension(f)
  File "/root/ufl/ufl/domain.py", line 376, in find_geometric_dimension
    error("Cannot determine geometric dimension from expression.")
  File "/root/ufl/ufl/log.py", line 135, in error
    raise self._exception_type(self._format_raw(*message))
ufl.log.UFLException: Cannot determine geometric dimension from expression.
```
Maybe I just need to tweak something in the code, but I don't know what. If you can share your example with UPML maybe I can spot what I am missing.